webExtensionApis/study/schema.yaml

```diff
@@ -0,0 +1,359 @@
+      - study:ready  OR
+      - study:endStudy
+
+      Preferences set
+      - `shield.${runtime.id}.firstRunTimestamp`
+
+      Note:
+      1. allowEnroll is ONLY used during first run (install)
+
+    parameters:
+    - name: studySetup
+      $ref: studySetup # TODO, for now, making progress up above
+
+  - name: endStudy
+    type: function
+    async: true
+    defaultReturn: {urls: ['url1','url2'], endingName:'some-reason'}
+    description: |
+      Signal to browser.study that it should end.
+
+      Usage scenarios:
+      - addons defined
+        - postive endings (tried feature)
+        - negative endings (client clicked 'no thanks')
+        - expiration / timeout (feature should last for 14 days then uninstall)
+
+      Logic:
+      - If study has already ended, do nothing.
+      - Else: END
+
+      END:
+      - record internally that study is ended.
+      - disable all methods that rely on configuration / setup.
+      - clear all prefs stored by `browser.study`
+      - fire telemetry pings for:
+        - 'exit'
+        - the ending, one of:
+
+          "ineligible",
+          "expired",
+          "user-disable",
+          "ended-positive",
+          "ended-neutral",
+          "ended-negative",
+
+      - augment all ending urls with query urls
+      - fire 'study:end' event to `browser.study.onEndStudy` handlers.
+
+      Addon should then do
+      - open returned urls
+      - feature specific cleanup
+      - uninstall the addon
+
+      Note:
+      1.  calling this function multiple time is safe.
+      `browser.study` will choose the
+
+    parameters:
+    - name: anEndingAlias
+      type: ['string', 'null']
+    - name: anEndingObject
+      type: object
+      optional: true
+    # should be anEnding Object
+
+  ## study / client information things
+  - name: info
+    type: function
+    async: true
+    description: |
+      current study configuration, including
+      - variation
+      - activeExperimentName
+      - timeUntilExpire
+      - firstRunTimestamp
+
+      But not:
+      - telemetry clientId
+
+      Throws Error if called before `browser.study.setup`
+
+    defaultReturn: {variation: "styleA", firstRunTimestamp: 1523968204184, activeExperimentName: "some experiment", timeUntilExpire: null}  # TODO add rest
+    parameters: []
+
+  - name: getDataPermissions
+    type: function
+    async: true
+    description: object of current dataPermissions with keys shield, pioneer, telemetry, 'ok'
+    defaultReturn: {shield: true, pioneer: false, telemetry: true, alwaysPrivateBrowsing: false}
+    parameters: []
+
+  # telemetry related things
+  - name: sendTelemetry
+    type: function
+    description: |
+      Send Telemetry using appropriate shield or pioneer methods.
+
+      shield:
+      - `shield-study-addon` ping, requires object string keys and string values
+
+      pioneer:
+      - TBD
+
+      Note:
+      - no conversions / coercion of data happens.
+
+      Note:
```

Comment:
Suggestion: Throw exceptions in these cases

Reply:
Yes, I agree on that.  Might be easier to say than do :)